pkg/httpmock/registry.go

```diff
@@ -25,6 +28,18 @@ func (r *Registry) Register(m Matcher, resp Responder) {
 	})
 }
 
+func (r *Registry) Exclude(t *testing.T, m Matcher) {
+	excludedStub := &Stub{
+		Matcher: m,
+		Responder: func(req *http.Request) (*http.Response, error) {
+			assert.FailNowf(t, "Exclude error", "API called when excluded: %v", req.URL)
+			return nil, nil
+		},
+		exclude: true,
+	}
+	r.stubs = append(r.stubs, excludedStub)
+}
+
```

Comment:
**How readily can this new capability be reasoned in context of `httpmock`?**

This is the question I mentally keep coming back to reading and re-reading through this updated `registry` code.  Depending how accurate I am below, maybe some added documentation and/or renaming makes sense:

- Tests that `Register` stubs are ensuring that specific HTTP requests are being made for REST or GraphQL APIs when the test calls `Verify` after being deferred.  It in turn iterates over all the stubs registered, finds whether any stubs have not been matched, and errors appropriately.

- `Exclude` now changes this dynamic by including stubs that cause the test to fail if encountered.  Less of a `Exclude` and more of a `FailIf` scenario in my mind.

- Lastly, I don't know if `exclude` field is actually needed if an `Exclude`ed stub will be marked as `matched` in `RoundTrip`:

https://github.com/cli/cli/blob/95b83872156a7dd2f16cd37ad7b29de2871a2d85/pkg/httpmock/registry.go#L64-L83

Reply:
> Lastly, I don't know if exclude field is actually needed if an Excludeed stub will be marked as matched in RoundTrip

Unfortunately this is required because `Verify` will complain for any unmatched stubs. 

There's a work-around for this by adding an `excludedStubs` key to the `Registry` struct, but I tinkered with that for a long time and the looping and exiting of the tests weren't very user friendly. There ended up being extra log dump to parse for the Exclusion failure that was bad UX and the `exclude` field is where I landed.

> Exclude now changes this dynamic by including stubs that cause the test to fail if encountered. Less of a Exclude and more of a FailIf scenario in my mind

First, names are hard 😅 I'd love to figure out what best to name this 📛 

That said, I think `Exclude` isn't bad because technically we're calling `registry.Exclude` when we use this - it's just shortened to `r.Exclude` in many of these tests (it can be confusing because we're shortening the name of the registry. Shortening the names of receivers and variables in golang is one of the things I take exception to in Golang practices for precisely this reason). So really our interface is:

```Golang
type Registry interface {
    Register(Matcher, Responder)
    Exclude(*testing.T, Matcher)
    Verify(Testing)
    RoundTrip(*http.Request)
}
```

Taken all together, you either "register" and api call "exclude" an api calls. 

I'm not sure if `FailIf` fits that well. I'm open to other suggestions.

Reply:
> First, names are hard 😅

Yes they are!

Putting aside whether to change the name of this or not, **I primarily wanted to make sure I was following what this intended**.

Because there is little to no documentation around the `httpmock` package, it causes the reader to read the code and infer the intention rather than clearly calling out what it does with documentation.  So I'm making sure I understand its use first 👍

Reply:
Ah, gotcha. Yeah, your understanding is correct, but I'll add one piece that I think will help solidify this

> Tests that Register stubs are ensuring that specific HTTP requests are being made for REST or GraphQL APIs when the test calls Verify after being deferred. It in turn iterates over all the stubs registered, finds whether any stubs have not been matched, and errors appropriately.

When a match is found (and before anything errors), it also runs the [provided callback, called `Responder`](https://github.com/cli/cli/blob/c81ccab4b8bb1ef1c774f4ba9851e9dcb11fa5e7/pkg/httpmock/registry.go#L77), on the stub. I've leveraged this so that the callback is hardcoded for the `Exclude` method so that when the match is found the callback that is invoked results in a failing test. That way I can avoid relying on errors to communicate that an excluded endpoint was called. This is important because testing for errors is a common pattern throughout the code (see [wantErr](https://github.com/cli/cli/blob/c81ccab4b8bb1ef1c774f4ba9851e9dcb11fa5e7/pkg/cmd/repo/create/create_test.go#L858-L862))